File: [scripts]/al_mdt/server/modules/characters.lua
function FetchCharactersByName(firstName, lastName, gender) -- Fetch characters by name and gender filter
    if firstName ~= nil and lastName ~= nil then
        local resp = DbQueries.characters.FetchByName(firstName, lastName, gender)

        for i=1, #resp do
            local char = resp[i]

            local finalResults = {
                identifier = char.identifier,
                firstName = char.firstName,
                lastName = char.lastName,
                dob = char.dob,
                sex = char.sex,
                photoId = '',
                contact = {
                    phone = '',
                    address = '',
                    relatedTo = {}
                },
                notes = '',
            }

            resp[i] = finalResults
        end

        return resp
    else
        return false
    end
end

function FetchCharacter(identifier) -- Fetch character
    if identifier ~= nil then
        local resp = DbQueries.characters.FetchByIdentifier(identifier)

        if resp then
            local char = resp

            char = {
                identifier = char.identifier,
                firstName = char.firstName,
                lastName = char.lastName,
                dob = char.dob,
                sex = char.sex,
                photoId = '',
                contact = {
                    phone = '',
                    address = '',
                    relatedTo = {}
                },
                notes = '',
            }

            return char
        else
            return false
        end

    else
        return false
    end
end

function FetchCharacterLicenses(identifier) -- Fetch character licenses
    return DbQueries.characters.FetchLicenses(identifier)
end
RegisterCallback('al_mdt:FetchCharacterLicenses', function(source, ...)
    TriggerProtFunction('FetchCharacterLicenses', source, function(src, cb, key, ...)
        if VerifyKey(src, key) then
            cb(FetchCharacterLicenses(...))
        end
    end, ...)
end)

RegisterCallback('al_mdt:FetchCharName', function(source, ...)
    TriggerProtFunction('FetchCharName', source, function(src, cb, key, ...)
        if VerifyKey(src, key) then
            cb(FetchCharName(...))
        end
    end, ...)
end)

function RevokeLicense(src, identifier, license) -- Revoke license
    if identifier ~= nil and license ~= nil then
        DbQueries.characters.RevokeLicense(identifier, license)
        RevokedLicenseAlert(src, identifier, license)
    end
end
RegisterNetEvent('al_mdt:RevokeLicense', function(...)
    TriggerProtFunction('RevokeLicense', source, function(src, key, ...)
        if VerifyKey(src, key) then
            RevokeLicense(src, ...)
        end
    end, ...)
end)

RegisterCallback('al_mdt:GrantLicense', function(source, ...) -- Grant license
    TriggerProtFunction('GrantLicense', source, function(src, cb, key, ...)
        if VerifyKey(src, key) then
            cb(GrantLicense(src, ...))
        end
    end, ...)
end)

File: [scripts]/al_mdt/ui/src/api/modules/characters.js
//import store from '@/store'
const resourceName = 'al_mdt'

class charactersApi {
    async Post(method, data) {
        if (process.env.NODE_ENV !== 'development') {
            const formattedData = data === undefined ? '{}' : JSON.stringify(data)

            try {
                let response = await fetch(`https://${resourceName}/${method}`, {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json; charset=UTF-8',
                    },
                    body: formattedData
                });

                return await response.json()
            } catch (error) {
                return false
            }
        }
    }

    // [POST] Fetch Data
    async GetCharacterByName(firstName, lastName, gender) {
        let charData = await this.Post('getCharacterByName', {
            firstName: firstName,
            lastName: lastName,
            gender: gender === 'NA' ? false : gender
        })
        let queryResults = [];

        if (Array.isArray(charData)) {
            charData.forEach((val, index, arry) => {
                if (val) {
                    if (gender !== "NA" && val.sex === gender) { // With gender filter
                        if (val.firstName.toLowerCase().match(firstName.toLowerCase())) {// First Name Match
                            if (val.lastName.toLowerCase().match(lastName.toLowerCase())) {// Last Name Match
                                queryResults.unshift(arry[index]);
                            } else {
                                queryResults.push(arry[index]);
                            }
                        } else if (val.lastName.toLowerCase().match(lastName.toLowerCase())) { // Last Name Match
                            queryResults.push(arry[index]);
                        }
                    } else if (gender === "NA") {// No gender filter
                        if (val.firstName.toLowerCase().match(firstName.toLowerCase())) {// First Name Match
                            if (val.lastName.toLowerCase().match(lastName.toLowerCase())) {// Last Name Match
                                queryResults.unshift(arry[index]);
                            } else {
                                queryResults.push(arry[index]);
                            }
                        } else if (val.lastName.toLowerCase().match(lastName.toLowerCase())) {// Last Name Match
                            queryResults.push(arry[index]);
                        }
                    }
                }
            });
        }
        return queryResults
    }

    // [POST] Fetch detailed profile
    async GetDetailedProfile(identifier) {
        return await this.Post('getCharacterDetailedProfile', { identifier: identifier })
    }

    // [POST] Fetch licenses
    async GetCharacterLicenses(identifier) {
        let licenses = await this.Post('getCharacterLicenses', identifier)

        return licenses
    }

    // [POST] Revoke license
    RevokeLicense(identifier, license) {
        this.Post('revokeLicense', {
            identifier: identifier,
            license: license
        })
    }

    // [POST] Grant license
    async GrantLicense(identifier, license) {
        return await this.Post('grantLicense', {
            identifier: identifier,
            license: license
        })
    }

    // [POST] Update profile
    UpdateProfile(identifier, photoId, contact, notes) {
        this.Post('updateProfile', {
            identifier: identifier,
            photoId: photoId,
            contact: contact,
            notes: notes
        })
    }
}

const instance = new charactersApi()

export default instance